src/state.rs

```diff
@@ -93,15 +99,42 @@ impl BarState {
     }
 
     pub(crate) fn set_message(&mut self, now: Instant, msg: Cow<'static, str>) {
-        self.state.message = msg;
+        self.state.message = TabExpandedString::new(msg, self.tab_width);
         self.update_estimate_and_draw(now);
     }
 
+    // Called in builder context
+    pub(crate) fn set_message_without_draw(&mut self, msg: Cow<'static, str>) {
+        self.state.message = TabExpandedString::new(msg, self.tab_width);
+    }
+
     pub(crate) fn set_prefix(&mut self, now: Instant, prefix: Cow<'static, str>) {
-        self.state.prefix = prefix;
+        self.state.prefix = TabExpandedString::new(prefix, self.tab_width);
+        self.update_estimate_and_draw(now);
+    }
+
+    // Called in builder context
+    pub(crate) fn set_prefix_without_draw(&mut self, prefix: Cow<'static, str>) {
+        self.state.prefix = TabExpandedString::new(prefix, self.tab_width);
+    }
+
+    pub(crate) fn set_tab_width(&mut self, now: Instant, tab_width: usize) {
```

Comment:
I don't think we need both with and without draw variants, let's only do the `without_draw()` variant and drop the suffix.